avalanche/models/packnet.py

```diff
@@ -0,0 +1,527 @@
+        self.register_buffer(
+            "visible_mask", torch.ones_like(self.weight, dtype=torch.bool)
+        )
+
+        self.unfrozen_mask: Tensor
+        """Mask of weights that are mutable"""
+        self.register_buffer(
+            "unfrozen_mask", torch.ones_like(self.weight, dtype=torch.bool)
+        )
+
+        self._state: Tensor
+        """The current state of the PackNet, see `State` for more information"""
+        self.register_buffer(
+            "_state", torch.tensor(self.State.TRAINING.value, dtype=torch.int)
+        )
+
+        self.weight.register_hook(self._remove_gradient_hook)
+
+    @property
+    @abstractmethod
+    def weight(self) -> Tensor:
+        raise NotImplementedError("weight not implemented")
+
+    @property
+    @abstractmethod
+    def bias(self) -> Tensor:
+        raise NotImplementedError("weight not implemented")
+
+    @property
+    def pruned_mask(self) -> Tensor:
+        """Return a mask of weights that have been pruned"""
+        return self.task_index.eq(self.PRUNED_CODE)
+
+    @property
+    def state(self) -> PackNetModule.State:
+        return self.State(self._state.item())
+
+    @state.setter
+    def state(self, state: PackNetModule.State):
+        self._state.fill_(state.value)
+
+    def prune(self, prune_proportion: float):
+        self._state_guard([self.State.TRAINING], self.State.POST_PRUNE)
+        ranked = self._rank_prunable()
+        prune_count = int(len(ranked) * prune_proportion)
+        self._prune_weights(ranked[:prune_count])
+        self.unfrozen_mask = self.task_index.eq(self._task_count)
+
+    def available_weights(self) -> Tensor:
+        return self.visible_mask * self.weight
+
+    def train_uncommitted(self):
+        self.activate_task(self.task_count())
+        self._state_guard([self.State.EVAL], self.State.TRAINING)
+
+        self.unfrozen_mask = self.task_index.eq(self.task_count())
+        self.visible_mask = self.visible_mask | self.unfrozen_mask
+
+    def activate_task(self, task_id: int):
+        self._state_guard(
+            [self.State.EVAL, self.State.TRAINING],
+            self.State.EVAL,
+        )
+        self.visible_mask.zero_()
+        self._is_subset_id_valid(task_id)
+        self.visible_mask = self.visible_mask | self.task_index.eq(task_id)
+
+    def freeze_weights(self):
+        self._state_guard([self.State.POST_PRUNE], self.State.EVAL)
+        self._task_count += 1
+        self.task_index[self.pruned_mask] = self._task_count.item()
+        # Change the active z_index
+        self.unfrozen_mask.zero_()
+
+        if self.bias is not None:
+            self.bias.requires_grad = False
+
+    def task_count(self) -> int:
+        return int(self._task_count.item())
+
+    @property
+    def device(self) -> torch.device:
+        return self.weight.device
+
+    def _remove_gradient_hook(self, grad: Tensor) -> Tensor:
+        """
+        Only the top layer should be trained. Todo so all other gradients
```

Comment:
Can we do it differently if we only need to backprop on the classifier? for example, `h.detach()` when we don't need the gradients? it would be more efficient.

Reply:
I don't believe this is possible, because we still need many of the gradients i.e `task_index` is the same shape as the grad/weights.